File: 01_spiralMatrix/1.py
class Solution:
    def spirallyTraverse(self, mat):
       
        m, n = len(mat), len(mat[0])
        ans = []
        top, bottom, left, right = 0, m - 1, 0, n - 1

        while top <= bottom and left <= right:

            for i in range(left, right + 1):
                ans.append(mat[top][i])
            top += 1

            for i in range(top, bottom + 1):
                ans.append(mat[i][right])
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    ans.append(mat[bottom][i])
                bottom -= 1

            if left <= right:
                for i in range(bottom, top - 1, -1):
                    ans.append(mat[i][left])
                left += 1

        return ans

sol = Solution()
mat = [
    [1,  2,  3,  4],
    [5,  6,  7,  8],
    [9, 10, 11, 12],
    [13,14,15,16]
]
print("Spiral Traversal:", sol.spirallyTraverse(mat))

File: 01_spiralMatrix/1.c
#include <stdio.h>

void spirallyTraverse(int rows, int cols, int mat[rows][cols]) {
    int top = 0, bottom = rows - 1;
    int left = 0, right = cols - 1;

    while (top <= bottom && left <= right) {

        for (int i = left; i <= right; i++) {
            printf("%d ", mat[top][i]);
        }
        top++;

        for (int i = top; i <= bottom; i++) {
            printf("%d ", mat[i][right]);
        }
        right--;

        if (top <= bottom) {
            for (int i = right; i >= left; i--) {
                printf("%d ", mat[bottom][i]);
            }
            bottom--;
        }

        if (left <= right) {
            for (int i = bottom; i >= top; i--) {
                printf("%d ", mat[i][left]);
            }
            left++;
        }
    }
}

int main() {
    int mat[4][4] = {
        {1,  2,  3,  4},
        {5,  6,  7,  8},
        {9, 10, 11, 12},
        {13,14, 15,16}
    };

    int rows = 4, cols = 4;

    printf("Spiral Traversal of the matrix:\n");
    spirallyTraverse(rows, cols, mat);

    return 0;
}